compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmStringConcatenationLowering.kt

```diff
@@ -0,0 +1,165 @@
+import org.jetbrains.kotlin.ir.declarations.IrSimpleFunction
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.expressions.IrStringConcatenation
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.util.constructors
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.ir.visitors.transformChildrenVoid
+import org.jetbrains.kotlin.name.Name
+
+internal val jvmStringConcatenationLowering = makeIrFilePhase(
+    ::JvmStringConcatenationLowering,
+    name = "StringConcatenation",
+    description = "Replace IrStringConcatenation with string builders"
+)
+
+/**
+ * This lowering pass replaces [IrStringConcatenation]s with StringBuilder appends.
+ *
+ * This pass is based on [StringConcatenationLowering] in backend.common. The main difference
+ * is that this pass also handles JVM specific optimizations, such as calling stringPlus
+ * for two arguments, and properly handles inline classes.
+ */
+private class JvmStringConcatenationLowering(val context: JvmBackendContext) : FileLoweringPass, IrElementTransformerVoidWithContext() {
+    override fun lower(irFile: IrFile) = irFile.transformChildrenVoid()
+
+    private val typesWithSpecialStringFunction = context.irBuiltIns.primitiveIrTypes + context.irBuiltIns.stringType
+
+    private val nameAppend = Name.identifier("append")
+    private val nameValueOf = Name.identifier("valueOf")
+
+    private val stringBuilder = context.ir.symbols.stringBuilder.owner
+
+    private val constructor = stringBuilder.constructors.single {
+        it.valueParameters.size == 0
+    }
+
+    private val IrClass.toStringFunction: IrSimpleFunction
+        get() = functions.single {
+            with(FlattenStringConcatenationLowering) { it.isToString }
+        }
+
+    private val toStringFunction = stringBuilder.toStringFunction
+
+    private val defaultAppendFunction = stringBuilder.functions.single {
+        it.name == nameAppend &&
+                it.valueParameters.size == 1 &&
+                it.valueParameters.single().type.isNullableAny()
+    }
+
+    private val appendFunctions: Map<IrType, IrSimpleFunction?> =
+        typesWithSpecialStringFunction.associateWith { type ->
+            stringBuilder.functions.singleOrNull {
+                it.name == nameAppend && it.valueParameters.singleOrNull()?.type == type
+            }
+        }
+
+    private fun typeToAppendFunction(type: IrType): IrSimpleFunction =
+        appendFunctions[type] ?: defaultAppendFunction
+
+    private val defaultValueOfFunction = context.ir.symbols.javaLangString.owner.functions.single {
+        it.name == nameValueOf && it.valueParameters.singleOrNull()?.type?.isNullableAny() == true
+    }
+
+    private val valueOfFunctions: Map<IrType, IrSimpleFunction?> =
+        typesWithSpecialStringFunction.associateWith { type ->
+            context.ir.symbols.javaLangString.owner.functions.singleOrNull {
```

Comment:
Currently we'll iterate over all `java.lang.String` members (of which there are ~100) 9 times for each source file. Maybe we could do this traversal only once per compilation? Basically store this info somewhere in `JvmSymbols`, or `JvmBackendContext`

Reply:
Done! I've moved the code in question to JvmSymbols.